cronitor/celery.py

```diff
@@ -0,0 +1,178 @@
+
+logger = logging.getLogger(__name__)
+try:
+    import celery
+    import celery.beat
+    from celery.schedules import crontab, schedule, solar
+    from celery.signals import beat_init, task_prerun, task_failure, task_success, task_retry
+
+    if typing.TYPE_CHECKING:
+        from typing import Dict, List, Union, Optional, Tuple
+        import billiard.einfo
+        from celery.worker.request import Request
+except ImportError:
+    logger.error("Cannot use the cronitor.celery module without celery installed")
+    sys.exit(1)
+
+# For the signals to properly register, they need to be top-level objects.
+# Since they are defined dynamically in initialize(), we have to declare them up top,
+# make them global, and override them.
+celerybeat_startup = None
+ping_monitor_before_task = None
+ping_monitor_on_success = None
+ping_monitor_on_failure = None
+ping_monitor_on_retry = None
+
+
+def get_headers_from_task(task):  # type: (celery.Task) -> Dict
+    headers = task.request.headers or {}
+    headers.update(task.request.properties.get('application_headers', {}))
+    return headers
+
+
+def initialize(app, celerybeat_only=False, api_key=None):  # type: (celery.Celery, bool, Optional[str]) -> None
+    if api_key:
+        cronitor.api_key = api_key
+
+    if celerybeat_only:
+        cronitor.celerybeat_only = True
+
+    global celerybeat_startup
+    global ping_monitor_before_task
+    global ping_monitor_on_success
+    global ping_monitor_on_failure
+    global ping_monitor_on_retry
+
+    def celerybeat_startup(sender, **kwargs):  # type: (celery.beat.Service, Dict) -> None
+        scheduler = sender.get_scheduler()  # type: celery.beat.Scheduler
+        schedules = scheduler.get_schedule()
+        monitors = []  # type: List[Dict[str, str]]
+
+        for name in schedules:
+            if name.startswith('celery.'):
+                continue
+            entry = schedules[name]  # type: celery.beat.ScheduleEntry
+
+            # ignore all celerybeat scheduled events with the Cronitor exclusion header
+            headers = entry.options.pop('headers', {})
+            if headers.get('x-cronitor-exclude') in (True, 'true', 'True'):
+                logger.info("celerybeat entry '{}' ignored per exclusion header".format(name))
+                continue
+
+            item = entry.schedule  # type: celery.schedules.schedule
+            if isinstance(item, crontab):
+                cronitor_schedule = ('{0._orig_minute} {0._orig_hour} {0._orig_day_of_week} {0._orig_day_of_month} '
+                                     '{0._orig_month_of_year}').format(item)
+            elif isinstance(item, schedule):
+                freq = item.run_every  # type: datetime.timedelta
+                cronitor_schedule = 'every ' + humanize.precisedelta(freq)
+            elif isinstance(item, solar):
+                # We don't support solar schedules
+                logger.warning("The cronitor-python celery module does not support "
```

Comment:
TIL about Solar Schedules

Reply:
I think it's insane Celery offers this and I wish it didn't